src/asm_templates.rs

```diff
@@ -0,0 +1,174 @@
+        Directives::PrintPointer,
+    ]
+    .iter()
+    {
+        result = preprocess_directive(&result, directive.clone());
+    }
+    result
+}
+
+fn preprocess_directive(asm: &str, directive: Directives) -> String {
+    let (asm_replaced, messages) = replace_directives(asm, directive);
+    let result = add_data_section_for_directive(&asm_replaced, directive, messages);
+    result
+}
+
+/// replace all occurrences of the directive with the corresponding assembly code
+fn replace_directives(asm: &str, directive: Directives) -> (String, Vec<String>) {
+    let mut result = asm.to_owned().clone();
+    let mut prints: Vec<String> = Vec::new();
+
+    let (command_prefix, regex, cell_name, prefix, suffix) = match directive.clone() {
+        Directives::Print => {
+            // regex: print("<message>")
+            let print_regex = Regex::new(r#"print\("[^"]*"\)"#).expect("Invalid regex");
+            (PRINT_PREFIX, print_regex, "PRINT", r#"print(""#, r#"")"#)
+        }
+        Directives::Revert => {
+            // regex: revert("<message>")
+            let revert_regex = Regex::new(r#"revert\("[^"]*"\)"#).expect("Invalid regex");
+            (
+                EXCEPTION_PREFIX,
+                revert_regex,
+                "REVERT",
+                r#"revert(""#,
+                r#"")"#,
+            )
+        }
+        Directives::PrintRegister => {
+            // regex: print(<src>)
+            let print_reg_regex = Regex::new(r#"print\([^"\))]+\)"#).expect("Invalid regex");
+            (
+                PRINT_REG_PREFIX,
+                print_reg_regex,
+                "PRINT_REG",
+                r#"print("#,
+                r#")"#,
+            )
+        }
+        Directives::PrintPointer => {
+            // regex: printPtr(<src>)
+            let print_ptr_regex = Regex::new(r#"printPtr\([^"\)]+\)"#).expect("Invalid regex");
+            (
+                PRINT_PTR_PREFIX,
+                print_ptr_regex,
+                "PRINT_PTR",
+                r#"printPtr("#,
+                r#")"#,
+            )
+        }
+    };
+
+    for (_, matched) in asm.match_indices(&regex) {
+        let arg = matched
+            .strip_prefix(&prefix)
+            .expect("Invalid text in directive")
+            .strip_suffix(&suffix)
+            .expect("Invalid text in directive");
+
+        if directive == Directives::PrintRegister || directive == Directives::PrintPointer {
+            // ignore any args
+            if prints.is_empty() {
+                prints.push("".to_owned());
+            }
+        } else {
+            if arg.len() > 32 - command_prefix.len() {
+                panic!("Message inside directive is too long: {}", arg);
+            }
+            prints.push(arg.to_owned());
+        }
+
+        let reference_var =
+            "@".to_owned() + cell_name + "_" + &(prints.len() - 1).to_string() + "_STRING";
+        let mut line = "add ".to_owned() + &reference_var + ", r0, r0";
+
```

Comment:
also in general for rust - try to limit the 'mut'.
more rusty approach would be:

let line = format!("add ...");

let line = match directive {
      Revert => { format!("{}\n ret.panic, line)}
      PrintRegister => { ... }
}